src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,109 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private int size;
+    private T[] elementData;
+
+    public ArrayList() {
+        elementData = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        resize();
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index != size) {
+            checkIndex(index);
+        }
+        resize();
+        System.arraycopy(elementData, index, elementData, index + 1, size - index);
+        elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        final T oldObject = elementData[index];
+        System.arraycopy(
+                elementData, index + 1, elementData, index, size - index - 1);
+        elementData[--size] = null;
+        resize();
+        return oldObject;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (elementData[i] == null && element != null) {
+                continue;
+            }
+            if (elementData[i] == element || elementData[i].equals(element)) {
+                return remove(i);
+            }
+        }
+        throw new NoSuchElementException("Can't remove non-existing element: " + element);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void resize() {
+        if (size >= elementData.length) {
+            T[] newValues = (T[]) new Object[size + (size >> 1)];
+            System.arraycopy(elementData, 0, newValues, 0, size);
+            elementData = newValues;
+        }
+    }
+
+    private int indexOf(T value) {
+        for (int i = 0; i < size; i++) {
+            if ((value == elementData[i]) || (value != null && value.equals(elementData[i]))) {
+                return i;
+            }
+        }
+        return -1;
+    }
+
+    private void checkIndex(int index) {
```

Comment:
naming suggestion, being more descriptive can help, it's unclear from name for what are we checking, try `checkInBounds()` or something like that